tests/integration/long/test_policies.py

```diff
@@ -65,3 +71,127 @@ def test_should_rethrow_on_unvailable_with_default_policy_if_cas(self):
+                backoff_policy=ConstantReconnectionPolicy(0.1),
+                max_concurrent=1,
+                scope=ShardConnectionBackoffScope.Host,
+            )
+        )
+
+    def test_limited_concurrency_2_connection_per_host(self):
+        self._test_backoff(
+            LimitedConcurrencyShardConnectionBackoffPolicy(
+                backoff_policy=ConstantReconnectionPolicy(0.1),
+                max_concurrent=1,
+                scope=ShardConnectionBackoffScope.Host,
+            )
+        )
+
+    def test_no_delay(self):
+        self._test_backoff(NoDelayShardConnectionBackoffPolicy())
+
+    def _test_backoff(self, shard_connection_backoff_policy: ShardConnectionBackoffPolicy):
+        backoff_policy = None
+        if isinstance(shard_connection_backoff_policy, LimitedConcurrencyShardConnectionBackoffPolicy):
+            backoff_policy = shard_connection_backoff_policy.backoff_policy
+
+        cluster = TestCluster(
+            shard_connection_backoff_policy=shard_connection_backoff_policy,
+            reconnection_policy=ConstantReconnectionPolicy(0),
+        )
+        session = cluster.connect()
+        sharding_info = get_sharding_info(session)
+
+        # even if backoff is set and there is no sharding info
+        # behavior should be the same as if there is no backoff policy
+        if not backoff_policy or not sharding_info:
+            time.sleep(2)
+            expected_connections = 1
+            if sharding_info:
+                expected_connections = sharding_info.shards_count
+            for host_id, connections_count in get_connections_per_host(session).items():
+                self.assertEqual(connections_count, expected_connections)
+            return
+
+        sleep_time = 0
+        schedule = backoff_policy.new_schedule()
+        # Calculate total time it will need to establish all connections
+        if shard_connection_backoff_policy.scope == ShardConnectionBackoffScope.Cluster:
+            for _ in session.hosts:
+                for _ in range(sharding_info.shards_count - 1):
+                    sleep_time += next(schedule)
+            sleep_time /= shard_connection_backoff_policy.max_concurrent
+        elif shard_connection_backoff_policy.scope == ShardConnectionBackoffScope.Host:
+            for _ in range(sharding_info.shards_count - 1):
+                sleep_time += next(schedule)
+            sleep_time /= shard_connection_backoff_policy.max_concurrent
+        else:
+            raise ValueError("Unknown scope {}".format(shard_connection_backoff_policy.scope))
+
+        time.sleep(sleep_time / 2)
+        self.assertFalse(
+            is_connection_filled(shard_connection_backoff_policy.scope, session, sharding_info.shards_count))
+        time.sleep(sleep_time / 2 + 1)
+        self.assertTrue(
+            is_connection_filled(shard_connection_backoff_policy.scope, session, sharding_info.shards_count))
```

Comment:
If policies accepted a scheduler instead of `Session`, then this test could avoid having any sleeps - we could artificially control the scheduler. Much more could be unit tested as well!

Reply:
done